apps/src/templates/progress/progressHelpers.js

```diff
@@ -194,3 +197,87 @@ export const processedLevel = level => {
       level.sublevels && level.sublevels.map(level => processedLevel(level))
   };
 };
+
+const getLevelResult = serverProgress => {
+  if (serverProgress.status === LevelStatus.locked) {
+    return TestResults.LOCKED_RESULT;
+  }
+  if (serverProgress.readonly_answers) {
+    return TestResults.READONLY_SUBMISSION_RESULT;
+  }
+  if (serverProgress.submitted) {
+    return TestResults.SUBMITTED_RESULT;
+  }
+
+  return serverProgress.result || TestResults.NO_TESTS_RUN;
+};
+
+/**
+ * Parse a level progress object that we get from the server using either
+ * /api/user_progress or /dashboardapi/section_level_progress into our
+ * canonical studentLevelProgressType shape.
+ * @param {object} serverObject A progress object from the server
+ * @returns {studentLevelProgressType} Our canonical progress shape
+ */
+export const levelProgressFromServer = serverObject => {
+  return {
+    status: serverObject.status || LevelStatus.not_tried,
+    result: getLevelResult(serverObject),
+    paired: serverObject.paired || false,
+    timeSpent: serverObject.time_spent || 0,
+    pages:
+      serverObject.pages_completed && serverObject.pages_completed.length > 1
+        ? serverObject.pages_completed.map(
+            pageResult =>
+              (pageResult && levelProgressFromResult(pageResult)) ||
+              levelProgressWithStatus(LevelStatus.not_tried)
+          )
+        : null
+  };
+};
+
+/**
+ * Given an object from the server with student progress data keyed by level ID,
+ * parse the progress data into our canonical studenLevelProgressType
+ * @param {{levelId:serverProgress}} serverStudentProgress
+ * @returns {{levelId:studentLevelProgressType}}
+ */
+export const processServerStudentProgress = serverStudentProgress => {
+  return _.mapValues(serverStudentProgress, progress =>
+    levelProgressFromServer(progress)
+  );
+};
+
+/**
+ * Given an object from the server with section progress data keyed by student
+ * ID and level ID, parse the progress data into our canonical
+ * studenLevelProgressType
+ * @param {{studenId:{levelId:serverProgress}}} serverSectionProgress
+ * @returns {{studenId:{levelId:studentLevelProgressType}}}
+ */
+export const processServerSectionProgress = serverSectionProgress => {
+  const studentProgress = _.mapValues(serverSectionProgress, student =>
+    processServerStudentProgress(student)
+  );
+  return studentProgress;
+};
+
+/**
+ * Create a studentLevelProgressType object with the provided status string
+ * @param {string} status
+ * @returns {studentLevelProgressType}
+ */
+export const levelProgressWithStatus = status => {
```

Comment:
Is this directly analogous to `levelProgressFromResult`?  If so, could we call this `levelProgressWithResult` for symmetry?

Side note for future: Our code will have lots of these functions that convert from one representation to another.  We should come up with a naming convention for such functions (e.g. `statusToLevelProgress`, `levelProgressFromStatus`, `convertStatusToLevelProgress`, ...)

Reply:
good call, and i agree with your side note.